Bug 1465974 Add option for installing datadog-agent

jklukas opened the pull request:
Had some conversation on IRC indicating that turning on Datadog for all clusters would likely be prohibitively expensive due to per-host pricing. The idea here is to make datadog-agent explicitly opt-in by requiring command-line argument `--install-datadog`.

haroldwoo commented on ansible/files/bootstrap/telemetry.sh:
```diff
@@ -137,6 +145,28 @@ sudo yum makecache
+        export DD_API_KEY="$(sudo aws s3 cp $TELEMETRY_CONF_BUCKET/credentials/datadog -)"
+        sudo bash -c "$(curl -sLS https://raw.githubusercontent.com/DataDog/datadog-agent/master/cmd/agent/install_script.sh)"
+    fi
+    if [ "$METRICS_PROVIDER" = "stackdriver" ] ; then
```
I don't think we're moving forward with stackdriver for metrics

jklukas replied:
So we're planning to keep Datadog for metrics longer term? Do we have that documented anywhere at this point?

haroldwoo replied:
I would talk to Brian Pitts about this.  I don't think the long term solution has been decided yet

jklukas replied:
Talked to Brian. Sounds like Datadog for the near or medium term, so I'm going to remove the stackdriver bits from this PR.

haroldwoo commented on ansible/files/bootstrap/telemetry.sh:
```diff
@@ -137,6 +145,28 @@ sudo yum makecache
+if [ "$METRICS_PROVIDER" = "stackdriver" ] || [ "$LOGGING_PROVIDER" = "stackdriver" ]; then
+    # Install GCP credentials for stackdriver; see https://cloud.google.com/logging/docs/agent/authorization
+    sudo mkdir -p /etc/google/auth/
+    sudo aws s3 cp $TELEMETRY_CONF_BUCKET/credentials/gcp.json /etc/google/auth/application_default_credentials.json
```
who will have access to this bucket?

haroldwoo replied:
Just checked and this bucket lives in our cloudservices-aws-dev account, so all developers have access to it.  A better practice would be to generate credentials(which we would have to do for the aws link gcp project anyway) and put them in a s3 bucket which only the EMR iam has access to.  This is annoying because we either have to host that s3 bucket in a separate aws account and set up cross account access, or add a rule to explicitly deny access to this bucket to the group that all developers belong to

haroldwoo commented on ansible/files/bootstrap/telemetry.sh:
```diff
@@ -137,6 +145,28 @@ sudo yum makecache
+fi
+if [ "$LOGGING_PROVIDER" = "stackdriver" ] ; then
+    # https://cloud.google.com/logging/docs/agent/installation
+    sudo bash -c "$(curl -sLS https://dl.google.com/cloudagents/install-logging-agent.sh)"
```
This likely requires additional setup and configuration, esp. if this runs in AWS.

jklukas replied:
That may be the case? The stackdriver docs specifically talk about installing on AWS and it sounds like simply providing creds and installing the agent is enough to get a default set of logs shipping, including syslog. My plan is to deploy this to a personal staging bucket once we have creds figured out and start testing what gets to GCP to discover if additional configuration is needed.

haroldwoo replied:
A manual step for linking the aws account is required.  A service account + credentials then need to be created and placed on host.  The agent(fluentd) is preconfigured to send syslog + other things, but spark logs on emr likely require addl configuration.  I have outlined some steps https://mana.mozilla.org/wiki/display/SVCOPS/Stackdriver+Logging+for+AWS but this is for ec2 instances running containers

jklukas commented:
Thanks to both of you, @haroldwoo and @whd for looking this over and commenting.

bpitts is out today, but I will follow up with him about plans for the future of metrics and monitoring.

If it's not too scary, it would indeed unblock me from further investigation if we could provision a DataDog API key and google application creds and upload them to the bucket.

I opened https://bugzilla.mozilla.org/show_bug.cgi?id=1487073 to request the GCP creds. Is opening the bug helpful? Should I open one for Datadog creds as well? As @haroldwoo mentioned, I may not be fulling comprehending the scope of what's necessary to get Stackdriver hooked up, so let me know if this is a significant amount of work to make those creds available.

jklukas commented on ansible/files/bootstrap/telemetry.sh:
```diff
@@ -209,6 +221,35 @@ rm RRO-3.2.1-el6.x86_64.tar.gz
+tags:
+$(describe_cluster '.Cluster.Tags[] | "- \(.Key):\(.Value)"')
+- cluster_id:${CLUSTER_ID}
+EOL
```
For an example cluster I spun up similar to an ATMO cluster, this heredoc expanded to:

```yaml
init_config:
instances:
  - spark_url: http://ip-172-31-33-18:8088
    cluster_name: klukas-metrics-test
spark_cluster_mode: spark_yarn_mode
tags:
- App:telemetry-analysis
- Environment:prod
- Owner:jklukas@mozilla.com
- Type:worker
- Application:telemetry-analysis-worker-instance
- Name:klukas-metrics-test
- cluster_id:j-330MAU776X1LH
```

jklukas commented:
This is now complete and ready to review. This should not cause any change in behavior for new clusters since the new behavior requires setting `--metrics-provider`.

jklukas commented:
You can hold off on review. I realized I missed some nuance from discussion in https://bugzilla.mozilla.org/show_bug.cgi?id=1465974#c10 so going to refactor to use StatsD more directly. Should land tomorrow.

jklukas commented:
This now has Spark configured to report to Statsd rather than using Datadog's integration. Ready for review.

jklukas commented on ansible/files/configuration/configuration.json:
```diff
@@ -43,6 +44,11 @@
+    "Classification": "spark-metrics",
+    "Properties": {
+      "*.sink.statsd.class": "org.apache.spark.metrics.sink.StatsdSink"
+    }
```
This configuration means that every Spark cluster we launch will now be running StatsdSink, but it should fail silently if no statsd agent is running on the host.

jklukas commented on ansible/files/configuration/configuration.json:
```diff
@@ -23,6 +23,7 @@
   }, {
     "Classification": "spark-defaults",
     "Properties": {
+      "spark.metrics.namespace": "spark",
```
We set namespace to "spark" so that Spark doesn't use the default namespace of applicationId, which is unique for every cluster.

By setting this to spark and not configuring a prefix for the StatsdSink, we end up with metric paths like:

```
  spark.driver.LiveListenerBus.queue.appStatus.listenerProcessingTime.mean_rate
  spark.1.executor.shuffleTotalBytesRead
```

haroldwoo commented on ansible/files/bootstrap/telemetry.sh:
```diff
@@ -209,6 +221,35 @@ rm RRO-3.2.1-el6.x86_64.tar.gz
+if [ "$METRICS_PROVIDER" = "datadog" ] ; then
+    # Pull down and decrypt Datadog API key; see https://github.com/mozilla/emr-bootstrap-spark/pull/485
+    DD_DIR=$(mktemp -d)
+    aws s3 cp $TELEMETRY_CONF_BUCKET/packages/sops-3.0.4-1.x86_64.rpm $DD_DIR/
```
make the sops rpm name a variable in case we ever need to update it

haroldwoo commented on ansible/files/bootstrap/telemetry.sh:
```diff
@@ -209,6 +221,35 @@ rm RRO-3.2.1-el6.x86_64.tar.gz
+    sudo yum -y localinstall $DD_DIR/sops-3.0.4-1.x86_64.rpm
+    export DD_API_KEY=$(/usr/local/bin/sops --decrypt --extract '["datadog_agent::api_key"]' $DD_DIR/emr.yaml)
+    # Install datadog-agent; see https://app.datadoghq.com/account/settings#agent/aws
+    bash -c "$(curl -sLS https://raw.githubusercontent.com/DataDog/datadog-agent/master/cmd/agent/install_script.sh)"
```
I prefer using the vendor's install because they can update it without us having to make changes. It would be a good idea to add some monitoring to ensure that the agent is actually installed/running afterwards.

https://docs.datadoghq.com/agent/basic_agent_usage/ubuntu/?tab=agentv6
Do we want to set DD_UPGRADE=true to install v6?

jklukas commented on ansible/files/bootstrap/telemetry.sh:
```diff
@@ -209,6 +221,35 @@ rm RRO-3.2.1-el6.x86_64.tar.gz
+
+tags:
+- cluster_name:${CLUSTER_NAME}
+- cluster_id:${CLUSTER_ID}
```
`App` and `Application` are already provided as tags on the cluster, at least in the ATMO case, so the next line here (calling describe-clusters) will produce App and Application tags mirroring what's set on the cluster.

Does that sound fine? Does that need to be better documented?

jklukas replied:
Added commentary about what tags to expect.

haroldwoo commented on ansible/files/bootstrap/telemetry.sh:
```diff
@@ -209,6 +221,35 @@ rm RRO-3.2.1-el6.x86_64.tar.gz
+tags:
+- cluster_name:${CLUSTER_NAME}
+- cluster_id:${CLUSTER_ID}
+$(describe_cluster '.Cluster.Tags[] | "- \(.Key):\(.Value)"')
```
you could save the api result in a json file and jq it

jklukas replied:
Done.

jklukas commented on ansible/files/bootstrap/telemetry.sh:
```diff
@@ -13,6 +13,17 @@ then
+describe_cluster() {
+    local query=$1
+    jq -r "$query" describe-cluster.json
+}
```
We only make a single describe-cluster call now.

haroldwoo replied:
nice

haroldwoo commented on ansible/files/bootstrap/telemetry.sh:
```diff
@@ -209,6 +221,35 @@ rm RRO-3.2.1-el6.x86_64.tar.gz
+EOF
+
+    sudo restart datadog-agent
+elif [ -n "$METRICS_PROVIDER" ]; then
```
should this just be else? In case the user submits empty string? Or should we use a different flag altogether e.g. --use-datadog-metrics so we don't have to arg check

jklukas replied:
This was an attempt to stick with existing convention for other options. For `--email` and `--efs-dns`, if the options aren't supplied, then we never set the corresponding variable and we use a `-n "$EMAIL"` check to see if we need to do anything.

I'm not too concerned about the possibility of a user accidentally passing empty string as the argument. It seems more likely that they'd pass it with no argument, in which case `shift` would fail or the next flag would get swallowed.

jklukas commented:
I've been testing these changes in stage, and realized a few minor fixups and one necessary change.

In particular, I was only partially correct about it being fine to enable `StatsdSink` on all clusters. An EMR 5.13.0 cluster behaves just fine with StatsdSink outputting to a nonexistent agent. However, `StatsdSink` doesn't exist in earlier spark versions, so a 5.11.1 cluster with StatsdSink configured enabled will throw errors on attempting to instantiate a spark context.

I've squashed all the previously reviewed commits and added one additional commit that factors out metrics configuration to a new file that we can leave empty if the metrics option is not enabled.

jklukas commented on ansible/files/bootstrap/telemetry.sh:
```diff
@@ -209,6 +224,52 @@ rm RRO-3.2.1-el6.x86_64.tar.gz
+    if [[ "$EMR_LABEL" =~ ^emr-5.* && "$MIN_EMR_LABEL" > "$EMR_LABEL" ]]; then
+        echo "ERROR: Incompatible options. StatsdSink is not present in $EMR_LABEL, but is required to support --metrics-provider=datadog; must specify $MIN_EMR_LABEL or later to enable metrics; terminating" 1>&2
+        # Cause the cluster to terminate with "Bootstrap failure"
+        exit 1
```
I'm choosing to `exit 1` here to force the cluster to terminate. I tested this case and I'm able to access the logs to see what went wrong.

It seems better to force the cluster down during bootstrap than to let it come up and have the user be faced with mysterious errors when they try to run jobs.

jklukas commented:
Marking this for review again to see if you have any objections to the new changes in cb8fce460355f53aacf1d8b2ae3193434b3a0f55.

This is now behaving as expected in stage, though we're not able to actually install the datadog-agent in stage due to the credentials, sops package, etc. existing only in the prod S3 bucket.

haroldwoo commented on ansible/files/bootstrap/telemetry.sh:
```diff
@@ -209,6 +224,52 @@ rm RRO-3.2.1-el6.x86_64.tar.gz
+
+    MIN_EMR_LABEL=emr-5.13.0
+    EMR_LABEL=$(describe_cluster .Cluster.ReleaseLabel)
+    if [[ "$EMR_LABEL" =~ ^emr-5.* && "$MIN_EMR_LABEL" > "$EMR_LABEL" ]]; then
```
My only comment is that this may fail if they choose to change the way the emr label version strings are constructed

jklukas commented on ansible/files/bootstrap/telemetry.sh:
```diff
@@ -209,6 +224,62 @@ rm RRO-3.2.1-el6.x86_64.tar.gz
+if (parse_version('${SPARK_VERSION}') < parse_version('${MIN_SPARK_VERSION}')):
+   exit(1)
+EOF
+    }
```
This now parses based on Spark version rather than emr label, which should be more stable. Also, we now use python `pkg_version.parse_version` rather than string comparison.

haroldwoo replied:
lgtm
